### PR TITLE
Add `has` trap for accurate `in` with stores and mutables

### LINKED ISSUE #1107: [feature] Stores don't have `in`/`has` trap
### Describe the bug

Feature request: It'd be nice for the `in` operator to be reactive (`has` and `deleteProperty` proxy traps). Example:
### Your Example Website or App

https://playground.solidjs.com/?hash=-210189711&version=1.4.1

### Steps to Reproduce the Bug or Issue

https://playground.solidjs.com/?hash=-210189711&version=1.4.1

### Expected behavior

The effect should run when `n` is deleted in the example

### Screenshots or Videos

_No response_

### Platform

- OS: [e.g. macOS, Windows, Linux]
- Browser: [e.g. Chrome, Safari, Firefox]
- Version: [e.g. 91.1]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by edemaine**
## Summary

Fixes #1107 by adding `has` trap to Stores and Mutables, so that code like `if ('prop' in store) store.prop()` is properly reactive. In Discord it's also been requested that `$TRACK in store` should return `true`, so I also added support for that (and other node types, to match `get` — notably missing `$NAME` though).

## How did you test this change?

I haven't tested this yet, sorry, in particular because I still can't build Solid and run tests locally. But hopefully it's still useful, at least as a proof of concept? Feel free to reimplement it...

## Update: read before merging

[As discussed in #reactivity channel](https://discord.com/channels/722131463138705510/751355413701591120/996345879621537792), this might need special handling for `delete`d properties. Depends how we want to interpret `setStore(..., undefined)`...

### CODE REVIEW

File: packages/solid/store/src/mutable.ts [hunk 1]
```diff
@@ -39,6 +39,13 @@ const proxyTraps: ProxyHandler<StoreNode> = {
       : value;
   },
 
+  has(target, property) {
+    if (property === $RAW || property === $PROXY || property === $TRACK ||
+        property === $NODE || property === "__proto__") return true;
+    getDataNodes(target)[property]?(); // track
```
Comment (TiagoCavalcante) [inline packages/solid/store/src/mutable.ts:45]: I guess it should be `?.`

**Comment by ryansolid** [comment]
Hmm.. `in` not tracking is something we've used to our advantage in the past I believe. I'd need to check. Also things like the JSX heuristic should take this into consideration perhaps.

File: packages/solid/store/src/mutable.ts [hunk 2]
```diff
@@ -42,7 +42,7 @@ const proxyTraps: ProxyHandler<StoreNode> = {
   has(target, property) {
     if (property === $RAW || property === $PROXY || property === $TRACK ||
         property === $NODE || property === "__proto__") return true;
-    getDataNodes(target)[property]?(); // track
+    getDataNodes(target)[property]?.(); // track
```
Comment (ryansolid) [inline packages/solid/store/src/mutable.ts:45]: In general, I've avoid optional chaining in the code base as we don't do any transforms on the code base and in some environments it die. Every time one slips in something breaks something. As annoying as it is, that's why all the others are written the way they are.

Comment (edemaine) [reply packages/solid/store/src/mutable.ts:45]: Oops, right; I forgot.

**Comment by edemaine** [comment]
> Hmm.. `in` not tracking is something we've used to our advantage in the past I believe. I'd need to check.

I would have expected `unwrap` or `untrack` would be the way to avoid tracking. But should definitely check!

In core, I see that `For`, `Index`, `Suspense`, `observable`, `from`, and `applyState` use `in`, but I don't think the second argument can ever be a Store (e.g. it's `props`, an observer/producer, or an array).

> Also things like the JSX heuristic should take this into consideration perhaps.

Could you elaborate? Does the JSX compiler output `in` tests in some cases? Or are you suggesting it should?

**Comment by ryansolid** [comment]
Im suggesting right now this would never be considered reactive according to our heuristic.

```js
<Show when={"value" in store}>...</Show>
```

So if `in` is we need to change it.

**Comment by ryansolid** [comment]
Yeah that is all fine, just means it's probably a 1.5 sort of thing.

**Comment by ryansolid** [comment]
Thank you

**Pull request merged**